js/packages/teams-ai/src/authentication/BotAuthentication.ts

```diff
@@ -0,0 +1,188 @@
+import {
+    DialogSet,
+    DialogState,
+    DialogTurnResult,
+    DialogTurnStatus,
+    OAuthPrompt,
+    OAuthPromptSettings
+} from 'botbuilder-dialogs';
+import { DefaultTurnState } from '../DefaultTurnStateManager';
+import { TurnState } from '../TurnState';
+import { Application } from '../Application';
+import {
+    ActivityTypes,
+    InvokeResponse,
+    MemoryStorage,
+    TeamsSSOTokenExchangeMiddleware,
+    TurnContext,
+    Storage,
+    verifyStateOperationName,
+    tokenExchangeOperationName
+} from 'botbuilder';
+import { TurnStateProperty } from '../TurnStateProperty';
+
+/**
+ * @internal
+ */
+interface OAuthPromptResult {
+    connectionName: string;
+    token: string;
+}
+
+/**
+ * @internal
+ */
+interface UserAuthState {
+    signedIn: boolean;
+    message?: string;
+}
+
+/**
+ * @internal
+ */
+export class BotAuthentication<TState extends TurnState = DefaultTurnState> {
+    private _oauthPrompt: OAuthPrompt;
+    private _storage: Storage;
+    private _userSignInHandler?: (context: TurnContext, state: TState) => Promise<void>;
+
+    public constructor(app: Application<TState>, oauthPromptSettings: OAuthPromptSettings, storage?: Storage) {
+        // Create OAuthPrompt
+        this._oauthPrompt = new OAuthPrompt('OAuthPrompt', oauthPromptSettings);
+
+        this._storage = storage || new MemoryStorage();
+
+        // Handles deduplication of token exchange event when using SSO with Bot Authentication
+        app.adapter.use(new TeamsSSOTokenExchangeMiddleware(this._storage, oauthPromptSettings.connectionName));
+
+        // Add application routes to handle OAuth callbacks
+        app.addRoute(
+            (context) =>
+                Promise.resolve(
+                    context.activity.type === ActivityTypes.Invoke && context.activity.name === verifyStateOperationName
+                ),
+            async (context, state) => {
+                await this.continueConversationalBotAuthFlow(context, state);
+            },
+            true
+        );
+        app.addRoute(
+            (context) =>
+                Promise.resolve(
+                    context.activity.type === ActivityTypes.Invoke &&
+                        context.activity.name === tokenExchangeOperationName
+                ),
+            async (context, state) => {
+                await this.continueConversationalBotAuthFlow(context, state);
+            },
+            true
+        );
+    }
+
+    public async authenticate(context: TurnContext, state: TState): Promise<string | undefined> {
+        // Get property names to use
+        const userAuthStatePropertyName = this.getUserAuthStatePropertyName(context);
+        const userDialogStatePropertyName = this.getUserDialogStatePropertyName(context);
+
+        // Save message if not signed in
+        if (!state.conversation.value[userAuthStatePropertyName]) {
+            state.conversation.value[userAuthStatePropertyName] = {
```

Comment:
`signedIn` property is removed. Bot sso flow always ends with an incoming `signIn/verifyState` activity, because of that the initial message the user sent needs to be saved and then restored.